javascript/ql/test/library-tests/frameworks/Spife/tests.ql

```diff
@@ -33,6 +33,15 @@ query predicate passingPositiveTests(string res, string expectation, InlineTest
   res = "PASSED" and
   t.hasPositiveTest(expectation) and
   (
+    expectation = "headerDefinition" and
```

Comment:
@erik-krogh Im curious, what would be the best way to combine these 4 query predicates so I dont need to repeat each test item?

Reply:
Not sure what the best is.  
You could have a predicate `hasResult(File f, int line, string expectation)`.   
Then you could have a single `query predicate()` where you checked that they match up with the `InlineTest`.  

Something like (untested):  
```CodeQL
query predicate test(string expectation, string msg) {
  exists(InlineTest t | 
    t.hasPositiveTest(expectation) and 
    hasResult(t.getFile(), t.getLocation().getStartLine(), expectation) and
    result = "PASS. TP"
  )
  or
  exists(InlineTest t | 
    t.hasNegativeTest(expectation) and 
    hasResult(t.getFile(), t.getLocation().getStartLine(), expectation) and
    result = "FAIL FP"
  )
  // .... 
}
```

Reply:
Thanks, is there a way to check the subtype of a `DataFlow::Node` given its type name? eg:

```
test("handler", "Http::RouteHandler") or
test("source", "RemoteFlowSource") or
...
```

So that in `test` we do something like:

```
predicate test(expectation, nodeType) {
     ...
     exists(DataFlow::Node n | n instanceof nodeType
     ...
}
```

I tried `exists(DataFlow::Node n | n.getAQlClass() = nodeType ...)` but that returns specfic types (eg `StandardRouteHandler`) instead of the abstract class (eg: `Http::RouteHandler`)?

Reply:
> Thanks, is there a way to check the subtype of a `DataFlow::Node` given its type name?

No, unfortunately not. 
And `getAQlClass()` is only suitable for debugging, it's not suitable for a test like this (the results might change for unrelated reasons).

Reply:
Thanks! I will keep the current version then